Update testing guideline & template (#1791)
vdaas/vald: docs/contributing/coding-style.md

```go
type args struct {
    host string
    port string
}

type field struct {
    timeout time.Duration
}

type want struct {
    err error
}

type test struct {
    name       string
    args       args
    field      field
    want       want
    checkFunc func(err error) error
}

defaultCheckFunc := func(w want, err error) error {
    if err != w.err {
        return errors.Errorf("got error: %v, want: %v", err, w.err)
    }
}

tests := []test {
    {

>         name: "send success when host and port are correct value",

Reviewer: ```suggestion
        name: "success send when host and port are correct value",
```

```go
type args struct {

>     addr string

Reviewer: ```suggestion
    host string
```

1. goleak option

   The generated test code will default use the [goleak](https://github.com/uber-go/goleak) library to test if there is any goroutine leak.
   Sometimes you may want to skip the detection; for example, Vald uses the [fastime](https://github.com/kpango/fastime) library, but the internal goroutine is not closed due to the needs of the library.
   To skip the goleak detection, we need to create the following variable to store the ignore function.

   ```go
   var (
       // goroutine leak is detected by `fastime`, but it should be ignored in the test because it is an external package.
       goleakIgnoreOptions = []goleak.Option{
           goleak.IgnoreTopFunction("github.com/kpango/fastime.(*fastime).StartTimerD.func1"),
       }
   )
   ```

   And use it in `VerifyNone()` in the test case.

   ```go
   for _, tc := range tests {
       test := tc
       t.Run(test.name, func(tt *testing.T) {
           // modify the following line
           defer goleak.VerifyNone(tt, goleakIgnoreOptions...)
   ```

>    In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignore the common goleak functions in Vald use case.

Reviewer: ```suggestion
   In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignore the common goleak functions.
```

>    In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignoring the common goleak functions in Vald use case.
>    In test implementation, we can simplily import the internal goleak package and ignore all the necessary goleak function by default.

Reviewer 2: ```suggestion
   In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignoring the common goleak functions in Vald use case.
   In test implementation, we can simply import the internal goleak package and ignore all the necessary goleak function by default.
```

Author: How about it?
```suggestion
   In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignoring the common goleak functions in Vald use case.
   In test implementation, we can easily import the internal goleak package and ignore all the necessary goleak function by default.
```

Reviewer 2: Thank you for your reply!
I think your suggestion is better! 👍

   ```go
   // import internal goleak package
   import (
       "github.com/vdaas/vald/internal/test/goleak"
   )

   ...
   for _, tc := range tests {
       test := tc
       t.Run(test.name, func(tt *testing.T) {
           // use internal goleak to ignore necessary function by default
           defer goleak.VerifyNone(tt, goleak.IgnoreCurrent())
   ```

1. goleak usage

   There are two methods for valid goroutine leak:

   1. Use `goleak.VerifyNone()` to validate it on each test cases.

>    2. Use `goleak.VerifyTestMain()` to validate it on each package.

Reviewer: ```suggestion
   1. Use `goleak.VerifyTestMain()` to validate it on each package.
```

>    By default in Vald, the goroutine leak validation is executed in each test case. e.g.

Author: ```suggestion
   By default, the goroutine leak validation is executed in each test case. e.g.
```

>    By default, the goroutine leak validation is executed in each test case. e.g.

Reviewer: ```suggestion
   By default, the goroutine leak validation is executed in each test case. E.g.,
```

   ```go
   for _, tc := range tests {
        test := tc
        t.Run(test.name, func(tt *testing.T) {
            defer goleak.VerifyNone(tt, goleak.IgnoreCurrent())
   ```

>    In some cases, it may not work as some cleanup process or asynchronous process remaining in the background, and validating it on the test case using `goleak.VerifyNone()` may cause a false alarm.

Reviewer: ```suggestion
   In some cases, it may not work as some cleanup or asynchronous processes remaining in the background, and validating it on the test case using `goleak.VerifyNone()` may cause a false alarm.
```

>    To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each packages, to avoid the goleak false alarm.

Author: ```suggestion
   To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each package, to avoid the goleak false alarm.
```

>    To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each package, to avoid the goleak false alarm.

Reviewer: ```suggestion
   To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each package.
```

   For example:

   ```go
   for _, tc := range tests {
        test := tc
        t.Run(test.name, func(tt *testing.T) {
            defer goleak.VerifyNone(tt, goleak.IgnoreCurrent())

            // insert your defer function here

>             defer func(w want, tt *testing.T) {
>             // implement your defer func logic
>             if err:= recover(); err != nil {
>                 // check the panic
>             }
>             }(test.want, tt)

Reviewer 2: ```suggestion
            defer func(w want, tt *testing.T) {
                // implement your defer func logic
                if err:= recover(); err != nil {
                    // check the panic
                }
            }(test.want, tt)
```

   In Vald, we suggest to customize the test implmentation from the generated test code to create the context and manage the lifecycle in every test.

   We can remove the context from the `args` struct and create the context in test execution code.

   e.g.

   ```go
    // we can remove the ctx from args list
    type args struct {
        // ctx context.Context
	}
    ...

    for _, tc := range tests {
	    test := tc
		t.Run(test.name, func(tt *testing.T) {
            // create context here
            ctx, cancel := context.WithCancel(context.Background())
            // handle the context lifecycle by canceling the context after the test is executed
			defer cancel()

            ...

            // use the context created above
            got, gotErr := someFunc(ctx)

            ...
   ```

1. Struct initialization

   By default, when testing the function of the struct, the target struct initialization is implmeneted by setting the date from the `fields` defined in the test case.
   This initialization method has few disadvatages:

>    1. when there are many fields in the struct, it is hard to set them all

Author: ```suggestion
   1. When there are many fields in the struct, it is hard to set them all
```

>    2. the default value is the zero value of the type, not the struct default value from the struct initialization function

Author: ```suggestion
   2. The default value is the zero value of the type, not the struct default value from the struct initialization function
```

           // generated test code
   ```

1. Context

   If the target function accepts context as an input argument, the test code generated will include it in `args`.

>    e.g. 

Reviewer: ```suggestion
   E.g. ,
```

   ```go
    type args struct {
        ctx context.Context
    }
    ...
   ```

>    But in test implementation, it is hard to manage the context lifecycle. We want to guarantee the context is closed after the test is executed, to avoid any missing termination of the process.

Reviewer: ```suggestion
   But in test implementation, it is hard to manage the context lifecycle.
   We want to guarantee the context is closed after each test is executed to avoid any missing termination of the process.
```

   In Vald, we suggest customizing the test implementation from the generated test code to create the context and manage the lifecycle in every test.

   We can remove the context from the `args` struct and create the context in the test execution code.

>    e.g.

Reviewer: ```suggestion
   E.g.,
```

Author: ```suggestion
   E.g. ,
```

   ```go
    // we can remove the ctx from args list
    type args struct {
        // ctx context.Context
    }
    ...

    for _, tc := range tests {
        test := tc
        t.Run(test.name, func(tt *testing.T) {
            // create context here
            ctx, cancel := context.WithCancel(context.Background())
            // handle the context lifecycle by canceling the context after the test is executed
            defer cancel()

            ...

            // use the context created above
            got, gotErr := someFunc(ctx)

            ...
   ```

1. Struct initialization

>    By default, when testing the function of the struct, the target struct initialization is implemented by setting the date from the `fields` defined in the test case.

Reviewer: ```suggestion
   By default, when testing the function of the struct, the target struct initialization is implemented by setting the data from the `fields` defined in the test case.
```

   This initialization method has a few disadvantages:

>    1. When there are many fields in the struct, it is hard to set them all

Reviewer: ```suggestion
   1. When there are many fields in the struct, it is hard to set them all.
```

>    2. The default value is the zero value of the type, not the struct default value from the struct initialization function

Reviewer: ```suggestion
   2. The default value is the zero value of the type, not the struct default value from the struct initialization function.
```

Author: ```suggestion
   1. The default value is the zero value of the type, not the struct default value from the struct initialization function.
```

>    By default, when testing the function of the struct, the target struct initialization is implemented by setting the date from the `fields` defined in the test case.
>    This initialization method has a few disadvatages:
> 
>    1. When there are many fields in the struct, it is hard to set them all
>    2. The default value is the zero value of the type, not the struct default value from the struct initialization function
> 
>    To resolve these problems, we can modify the test implementation to use the struct initialization function instead of setting the struct fields on the test cases.
> 
>    For example, the current implementation may look like this:

Reviewer 2: ```suggestion
   By default, when testing the function of the struct, the target struct initialization is implemented by setting the date from the `fields` defined in the test case.
   This initialization method has a few disadvantages:

   1. When there are many fields in the struct, it is hard to set them all
   2. The default value is the zero value of the type, not the struct default value from the struct initialization function

   To resolve these problems, we can modify the test implementation to use the struct initialization function instead of setting the struct fields on the test cases.

   For example, the current implementation may look like this:
```

   ```go
   func Test_server_SaveIndex(t *testing.T) {
        ...

        // all the fields to initialize target struct
        type fields struct {
            name              string
            ip                string
            ngt               service.NGT
            eg                errgroup.Group
            streamConcurrency int
        }

        ...

        for _, tc := range tests {
            test := tc
            t.Run(test.name, func(tt *testing.T) {
                tt.Parallel()
                ...

>                 // we initialization the target struct using fields defined in test case

Author: ```suggestion
                // we initialize the target struct using fields defined in test case
```

        }

        ...

        for _, tc := range tests {
            test := tc
            t.Run(test.name, func(tt *testing.T) {
                tt.Parallel()
                ...

                // initialize required fields of server struct
                eg, _ := errgroup.New(ctx)
                ngt, err := service.New(test.fields.svcCfg, append(test.fields.svcOpts, service.WithErrGroup(eg))...)
                if err != nil {
                    tt.Errorf("failed to init ngt service, error = %v", err)
                }

                // we use the struct initialization function instead of setting the fields,
                // by combining the use of the functional option, we can configure the struct,
                // and the default values of the struct will be set automatically if we are not defining it.
                s, err := New(append(test.fields.srvOpts, WithNGT(ngt), WithErrGroup(eg))...)
                if err != nil {
                    tt.Errorf("failed to init server, error= %v", err)
                }

                gotRes, err := s.SaveIndex(test.args.ctx, test.args.in1)
                ...
   ```

### Parallel test

> In Vald, we use parallel tests to accelerate the execution of tests by default. There are two layers of enabling parallel tests.

Reviewer: ```suggestion
In Vald, we use parallel tests to accelerate the execution of tests by default.
There are two layers of enabling parallel tests.
```

1. Parallel for the test function

> 2. Parallel for the sub-tests in test function

Author: ```suggestion
2. Parallel for the subtests in test function
```

The generated test case will enable these 2 parallel mode by default. It is implemented by:

```go
func Test_server_CreateIndex(t *testing.T) {
    t.Parallel() // parallel for the test function
    type args struct {
   
    ...

    for _, tc := range tests {
        test := tc
        t.Run(test.name, func(tt *testing.T) {

>             tt.Parallel() // parallel for sub-tests

Author: ```suggestion
            tt.Parallel() // parallel for subtests
```

           // generated test code
   ```

1. Context

   If the target function accepts context as an input argument, the test code generated will include it in `args`.

>    E.g. ,

Author: removed `E.g.` here

   ```go
    type args struct {
        ctx context.Context
    }
    ...
   ```

   But in test implementation, it is hard to manage the context lifecycle.
   We want to guarantee the context is closed after each test is executed to avoid any missing termination of the process.

   In Vald, we suggest customizing the test implementation from the generated test code to create the context and manage the lifecycle in every test.

   We can remove the context from the `args` struct and create the context in the test execution code.

>    E.g. ,

Author: removed E.g. here

                tt.Parallel()
                ...

                // initialize required fields of server struct
                eg, _ := errgroup.New(ctx)
                ngt, err := service.New(test.fields.svcCfg, append(test.fields.svcOpts, service.WithErrGroup(eg))...)
                if err != nil {
                    tt.Errorf("failed to init ngt service, error = %v", err)
                }

                // we use the struct initialization function instead of setting the fields,
                // by combining the use of the functional option, we can configure the struct,
                // and the default values of the struct will be set automatically if we are not defining it.
                s, err := New(append(test.fields.srvOpts, WithNGT(ngt), WithErrGroup(eg))...)
                if err != nil {
                    tt.Errorf("failed to init server, error= %v", err)
                }

                gotRes, err := s.SaveIndex(test.args.ctx, test.args.in1)
                ...
   ```

### Parallel test

In Vald, we use parallel tests to accelerate the execution of tests by default.
There are two layers of enabling parallel tests.

1. Parallel for the test function

> 2. Parallel for the subtests in the test function

Reviewer: ```suggestion
1. Parallel for the subtests in the test function
```

                ...
   ```

### Parallel test

In Vald, we use parallel tests to accelerate the execution of tests by default.
There are two layers of enabling parallel tests.

1. Parallel for the test function
2. Parallel for the subtests in the test function

The generated test case will enable these two parallel modes by default. It is implemented by:

```go
func Test_server_CreateIndex(t *testing.T) {
    t.Parallel() // parallel for the test function
    type args struct {
   
    ...

    for _, tc := range tests {
        test := tc
        t.Run(test.name, func(tt *testing.T) {
            tt.Parallel() // parallel for subtests
    ...
```

Be careful of using the parallel test, avoid using share object used in the test, to avoid race detector to detect the race in test and fail the test.

### Helper function

In golang testing package, they defined the helper functions in testing types to indicate the current function is a test helper function and skip printing the file and line information to the output.

In Vald, we can apply it to the different helper functions like `beforeFunc()` or `afterFunc()` defined in the test case struct.

```go

> 

Author: ```suggestion
```
